### ISSUE TITLE
calculate the bounding box issue, how to correctly update the margin of the boulding box

### ISSUE THREAD

**Issue by asmwarrior** [GitHubLionel/wxMathPlot#22]
Hi, we have discussed in this long issue, especially in this comment, I draw an image diagram:

https://github.com/GitHubLionel/wxMathPlot/issues/20#issuecomment-2118541452

I'm not sure it is finished yet, so I start a new issue, so it won't got forgot. Thanks.

**Comment by GitHubLionel** [comment]
Hi,
Well I don't understand what you propose. Margins are defined constant in "General" page of configuration.
Change this may have several consequences since all depend of this values.
We can add an extra margin to the margin box simply in the mpWindow::OnPaint method :
On line 3255: `trgDc->DrawRectangle(m_margin.left-10, m_margin.top-10, m_plotWidth+20, m_plotHeight+20);`
In this way, no impact on the code.

**Comment by asmwarrior** [comment]
Sorry for the confusion, see the image below:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/631c589f-6383-4f9c-8025-e76e7391e0ca)

You see, the red arrow were pointing to the margins.

What I need to keep is the "gap" pointed by the blue arrows. Because after I "fit" the screen, I see the markers are not shown correctly in those edges.

**Comment by asmwarrior** [comment]
Another issue about calculating the bbox is that I reported here:

https://github.com/GitHubLionel/wxMathPlot/issues/20#issuecomment-2117620865

I believe the "delta" method is not correct.

I mean in the code is not correct. If you want to "keep some gap", you have to some percentage of the current actual bbox. That's why I wrote an image in the comment: https://github.com/GitHubLionel/wxMathPlot/issues/20#issuecomment-2118541452

~~~~
void mpFXYVector::SetData(const std::vector<double> &xs, const std::vector<double> &ys)
{
  // Check if the data vectora are of the same size
  if (xs.size() != ys.size())
  {
    wxLogError(_T("wxMathPlot error: X and Y vector are not of the same length!"));
    return;
  }
  // Copy the data:
  m_xs = xs;
  m_ys = ys;

  // Update internal variables for the bounding box.
  if (xs.size() > 0)
  {
    m_minX = m_maxX = xs[0];
    m_minY = m_maxY = ys[0];

    std::vector<double>::const_iterator it;

    // X scale
    it = xs.begin();
    m_lastX = (*it);
    m_deltaX = 1e+308; // Big number
    it++;
    for (; it != xs.end(); it++)
    {
      if (abs((*it) - m_lastX) < m_deltaX)
        m_deltaX = abs((*it) - m_lastX);
      m_lastX = (*it);

      if (*it < m_minX)
        m_minX = (*it) - m_deltaX;
      else
        if (*it > m_maxX)
          m_maxX = (*it) + m_deltaX;
    }

    // Y scale
    it = ys.begin();
    m_lastY = (*it);
    m_deltaY = 1e+308; // Big number
    it++;
    for (; it != ys.end(); it++)
    {
      if (abs((*it) - m_lastY) < m_deltaY)
        m_deltaY = abs((*it) - m_lastY);
      m_lastY = (*it);

      if (*it < m_minY)
        m_minY = (*it) - m_deltaY;
      else
        if (*it > m_maxY)
          m_maxY = (*it) + m_deltaY;
    }
  }
  else
  {
    m_minX = -1;
    m_maxX = 1;
    m_minY = -1;
    m_maxY = 1;
  }
  Rewind();
}
~~~~

**Comment by GitHubLionel** [comment]
No, computation of BBox is ok. 
This is not a problem of bounding computation. This is just a problem of border. We only want to overflow when we have a square or a circle.
I made a try with adding an EXTRA_MARGIN.
See in sandbox.

**Comment by asmwarrior** [comment]
OK, I'm testing the sandbox code, and here is the result image:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/125215e6-e98c-4df3-91b7-fcde259b08e0)

The markers are shown correctly now when they are near the margin.

I will test other problem I reported in previous comments.

**Comment by asmwarrior** [comment]
![margin-wxMathplot](https://github.com/GitHubLionel/wxMathPlot/assets/561818/0471b84a-213e-4b50-b5b1-d71c67b02201)

I just read the code, and see that the bounding box is calculated like the above image, am I correct?

I see that the `delta` is calculated by the `latest point - last point`, so in the above image, delta is `p4-p3`, then the calculated bounding box is extend by the `delta` if the current point(`p4`) is beyond the last bounding box.

I think this method is reasonable OK.

My idea in the [#20 (comment)](https://github.com/GitHubLionel/wxMathPlot/issues/20#issuecomment-2118541452) was using another method, I mean to keep 10% extended of the actual bounding box.

**Comment by GitHubLionel** [comment]
Ok I push on the master branch.
I profit to add a functionality : when you click on series name on legend, you go directly to it's configuration.

**Comment by asmwarrior** [comment]
> Ok I push on the master branch. I profit to add a functionality : when you click on series name on legend, you go directly to it's configuration.

Hi, good feature, thanks.

When running the latest git head code, I see a bug here, see the image show below:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/d0d972fe-1236-483b-98e2-d9d05fd29507)

When I hover the mouse on the legend window, the track window still shows the coordinates of the current X, Y, and Y2. I think it should not be there if the mouse is move to the "margin field".

Oh, wait, I see that the mouse coordinates is shown in all the "margin field", see below:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/98f7a1b7-f75d-4c6b-bc58-dff5a14a55ad)

So, I think this is not a bug, you don't need to fix it.

**Comment by GitHubLionel** [comment]
Yes it is not a bug but it is true that I have asking me about that. Maybe it will be better to restrict to the plot area ?
Or add a property "m_drawOutsideMargins" ?
Your opinion ?

**Comment by asmwarrior** [comment]
> Yes it is not a bug but it is true that I have asking me about that. Maybe it will be better to restrict to the plot area ? Or add a property "m_drawOutsideMargins" ? Your opinion ?

It is good to have an option to toggle the feature, thanks.

**Comment by GitHubLionel** [comment]
Ok I do that and I add a line page on configuration.

**Comment by asmwarrior** [comment]
Thanks, I have just tested the latest git head code. It works perfectly. Great job for the improvements.

**Comment by GitHubLionel** [comment]
Thanks :+1:
